src/runtime/runtime_atsame5x.go

```diff
@@ -0,0 +1,338 @@
+	}
+
+	// Enable DFLL48M clock
+	sam.OSCCTRL.DFLLCTRLA.Set(0)
+	sam.OSCCTRL.DFLLMUL.Set((0x1 << sam.OSCCTRL_DFLLMUL_CSTEP_Pos) |
+		(0x1 << sam.OSCCTRL_DFLLMUL_FSTEP_Pos) |
+		(0x0 << sam.OSCCTRL_DFLLMUL_MUL_Pos))
+	for sam.OSCCTRL.DFLLSYNC.HasBits(sam.OSCCTRL_DFLLSYNC_DFLLMUL) {
+	}
+
+	sam.OSCCTRL.DFLLCTRLB.Set(0)
+	for sam.OSCCTRL.DFLLSYNC.HasBits(sam.OSCCTRL_DFLLSYNC_DFLLCTRLB) {
+	}
+
+	sam.OSCCTRL.DFLLCTRLA.SetBits(sam.OSCCTRL_DFLLCTRLA_ENABLE)
+	for sam.OSCCTRL.DFLLSYNC.HasBits(sam.OSCCTRL_DFLLSYNC_ENABLE) {
+	}
+
+	sam.OSCCTRL.DFLLVAL.Set(sam.OSCCTRL.DFLLVAL.Get())
+	for sam.OSCCTRL.DFLLSYNC.HasBits(sam.OSCCTRL_DFLLSYNC_DFLLVAL) {
+	}
+
+	sam.OSCCTRL.DFLLCTRLB.Set(sam.OSCCTRL_DFLLCTRLB_WAITLOCK |
+		sam.OSCCTRL_DFLLCTRLB_CCDIS |
+		sam.OSCCTRL_DFLLCTRLB_USBCRM)
+	for !sam.OSCCTRL.STATUS.HasBits(sam.OSCCTRL_STATUS_DFLLRDY) {
+	}
+
+	// set GCLK7 to run at 2MHz, using DFLL48M as clock source
+	// GCLK7 = 48MHz / 24 = 2MHz
+	sam.GCLK.GENCTRL[7].Set((sam.GCLK_GENCTRL_SRC_DFLL << sam.GCLK_GENCTRL_SRC_Pos) |
+		(24 << sam.GCLK_GENCTRL_DIV_Pos) |
+		sam.GCLK_GENCTRL_GENEN)
+	for sam.GCLK.SYNCBUSY.HasBits(sam.GCLK_SYNCBUSY_GENCTRL_GCLK7) {
+	}
+
+	// Set up the PLLs
+
+	// Set PLL0 to run at 120MHz, using GCLK7 as clock source
+	sam.GCLK.PCHCTRL[1].Set(sam.GCLK_PCHCTRL_CHEN |
+		(sam.GCLK_PCHCTRL_GEN_GCLK7 << sam.GCLK_PCHCTRL_GEN_Pos))
+
+	// multiplier = 59 + 1 + (0/32) = 60
+	// PLL0 = 2MHz * 60 = 120MHz
+	sam.OSCCTRL.DPLL[0].DPLLRATIO.Set((0x0 << sam.OSCCTRL_DPLL_DPLLRATIO_LDRFRAC_Pos) |
+		(59 << sam.OSCCTRL_DPLL_DPLLRATIO_LDR_Pos))
+	for sam.OSCCTRL.DPLL[0].DPLLSYNCBUSY.HasBits(sam.OSCCTRL_DPLL_DPLLSYNCBUSY_DPLLRATIO) {
+	}
+
+	// MUST USE LBYPASS DUE TO BUG IN REV A OF SAMD51, via Adafruit lib.
+	sam.OSCCTRL.DPLL[0].DPLLCTRLB.Set((sam.OSCCTRL_DPLL_DPLLCTRLB_REFCLK_GCLK << sam.OSCCTRL_DPLL_DPLLCTRLB_REFCLK_Pos) |
+		sam.OSCCTRL_DPLL_DPLLCTRLB_LBYPASS)
+
+	sam.OSCCTRL.DPLL[0].DPLLCTRLA.Set(sam.OSCCTRL_DPLL_DPLLCTRLA_ENABLE)
+	for !sam.OSCCTRL.DPLL[0].DPLLSTATUS.HasBits(sam.OSCCTRL_DPLL_DPLLSTATUS_CLKRDY) ||
+		!sam.OSCCTRL.DPLL[0].DPLLSTATUS.HasBits(sam.OSCCTRL_DPLL_DPLLSTATUS_LOCK) {
+	}
+
+	// Set PLL1 to run at 100MHz, using GCLK7 as clock source
+	sam.GCLK.PCHCTRL[2].Set(sam.GCLK_PCHCTRL_CHEN |
+		(sam.GCLK_PCHCTRL_GEN_GCLK7 << sam.GCLK_PCHCTRL_GEN_Pos))
+
+	// multiplier = 49 + 1 + (0/32) = 50
+	// PLL1 = 2MHz * 50 = 100MHz
+	sam.OSCCTRL.DPLL[1].DPLLRATIO.Set((0x0 << sam.OSCCTRL_DPLL_DPLLRATIO_LDRFRAC_Pos) |
+		(49 << sam.OSCCTRL_DPLL_DPLLRATIO_LDR_Pos))
+	for sam.OSCCTRL.DPLL[1].DPLLSYNCBUSY.HasBits(sam.OSCCTRL_DPLL_DPLLSYNCBUSY_DPLLRATIO) {
+	}
+
+	// // MUST USE LBYPASS DUE TO BUG IN REV A OF SAMD51
+	sam.OSCCTRL.DPLL[1].DPLLCTRLB.Set((sam.OSCCTRL_DPLL_DPLLCTRLB_REFCLK_GCLK << sam.OSCCTRL_DPLL_DPLLCTRLB_REFCLK_Pos) |
+		sam.OSCCTRL_DPLL_DPLLCTRLB_LBYPASS)
+
+	sam.OSCCTRL.DPLL[1].DPLLCTRLA.Set(sam.OSCCTRL_DPLL_DPLLCTRLA_ENABLE)
+	// for !sam.OSCCTRL.DPLLSTATUS1.HasBits(sam.OSCCTRL_DPLLSTATUS_CLKRDY) ||
```

Comment:
Should the commented out section be removed?

Reply:
Right now, I'm working on runtime_atsamd51.go so that the differences are minimal.
I'd like to leave it as it is without any modifications in this PR.

```diff
$ diff -us src/runtime/runtime_atsamd51.go src/runtime/runtime_atsame5x.go
--- src/runtime/runtime_atsamd51.go     2021-04-15 09:13:27.819001300 +0900
+++ src/runtime/runtime_atsame5x.go     2021-04-15 09:13:27.819001300 +0900
@@ -1,4 +1,4 @@
-// +build sam,atsamd51
+// +build sam,atsame5x

 package runtime


```

Reply:
:+1: